taichi/program/async_engine.h

```diff
@@ -183,11 +179,26 @@ class AsyncEngine {
   void synchronize();
 
  private:
+  struct KernelMeta {
+    std::unique_ptr<Block> dummy_root;
+    std::vector<uint64> offloaded_hashes;
```

Comment:
This is a bit confusing to me: Why do offload hashes belong to the kernel meta? Is the same offload's hash different when the kernel is different?
I thought it should belong to the offloaded meta but the key of `offloaded_metas_` is the hash...

Reply:
> Why do offload hashes belong to the kernel meta?

Each offloaded task in a kernel will generate a hash. What I'm trying to avoid is the re-computation of the hashes for the offloads inside a kernel...

> Is the same offload's hash different when the kernel is different?

I don't think so. From what I can tell, the `hash()` function computes the hash purely based on the input IR, not the kernel info. So same offloads will generate same hashes... 

However, this identical relationship is kind of shaky. The current impl of `hash()` uses `irpass::print()` to get a textual AST `std::string`, then computes the hash of that string. If the `print()` ever mixes in the kernel info (e.g. prints the kernel name), then same offloaded tasks would result in different hashes.